File: web/src/generator/grammar.ts
import { Term as T } from "../lr/constants";
import { GenError } from "./error";

const TermFlag = {
	// This term is a terminal
	Terminal: 1,
	// This is the top production
	Top: 2,
	// This represents end-of-file
	Eof: 4,
	// This should be preserved, even if it doesn't occur in any rule
	Preserve: 8,
	// Rules used for * and + constructs
	Repeated: 16,
	// Rules explicitly marked as [inline]
	Inline: 32,
} as const;
type TermFlag = (typeof TermFlag)[keyof typeof TermFlag];

export type Props = { [name: string]: string };

export function hasProps(props: Props): boolean {
	for (let _p in props) return true;
	return false;
}

let termHash = 0;

export class Term {
	hash: number = ++termHash; // Used for sorting and hashing during parser generation
	id = -1; // Assigned in a later stage, used in actual output
	// Filled in only after the rules are simplified, used in automaton.ts
	rules: Rule[] = [];

	readonly name: string;
	private flags: number;
	readonly nodeName: string | null;
	readonly props: Props;

	constructor(name: string, flags: number, nodeName: string | null, props: Props = {}) {
		this.name = name;
		this.flags = flags;
		this.nodeName = nodeName;
		this.props = props;
	}

	toString(): string {
		return this.name;
	}
	get nodeType(): boolean {
		return this.top || this.nodeName != null || hasProps(this.props) || this.repeated;
	}
	get terminal(): boolean {
		return (this.flags & TermFlag.Terminal) > 0;
	}
	get eof(): boolean {
		return (this.flags & TermFlag.Eof) > 0;
	}
	get error(): boolean {
		return "error" in this.props;
	}
	get top(): boolean {
		return (this.flags & TermFlag.Top) > 0;
	}
	get interesting(): boolean {
		return this.flags > 0 || this.nodeName != null;
	}
	get repeated(): boolean {
		return (this.flags & TermFlag.Repeated) > 0;
	}
	set preserve(value: boolean) {
		this.flags = value ? this.flags | TermFlag.Preserve : this.flags & ~TermFlag.Preserve;
	}
	get preserve() {
		return (this.flags & TermFlag.Preserve) > 0;
	}
	set inline(value: boolean) {
		this.flags = value ? this.flags | TermFlag.Inline : this.flags & ~TermFlag.Inline;
	}
	get inline() {
		return (this.flags & TermFlag.Inline) > 0;
	}
	cmp(other: Term): number {
		return this.hash - other.hash;
	}
}

export class TermSet {
	terms: Term[] = [];
	// Map from term names to Term instances
	names: { [name: string]: Term } = Object.create(null);
	eof: Term;
	error: Term;
	tops: Term[] = [];

	constructor() {
		this.eof = this.term("␄", null, TermFlag.Terminal | TermFlag.Eof);
		this.error = this.term("⚠", "⚠", TermFlag.Preserve);
	}

	term(name: string, nodeName: string | null, flags: number = 0, props: Props = {}): Term {
		let term = new Term(name, flags, nodeName, props);
		this.terms.push(term);
		this.names[name] = term;
		return term;
	}

	makeTop(nodeName: string | null, props: Props): Term {
		const term = this.term("@top", nodeName, TermFlag.Top, props);
		this.tops.push(term);
		return term;
	}

	makeTerminal(name: string, nodeName: string | null, props = {}): Term {
		return this.term(name, nodeName, TermFlag.Terminal, props);
	}

	makeNonTerminal(name: string, nodeName: string | null, props = {}): Term {
		return this.term(name, nodeName, 0, props);
	}

	makeRepeat(name: string): Term {
		return this.term(name, null, TermFlag.Repeated);
	}

	uniqueName(name: string): string {
		for (let i = 0; ; i++) {
			let cur = i ? `${name}-${i}` : name;
			if (!this.names[cur]) return cur;
		}
	}

	finish(rules: readonly Rule[]): {
		nodeTypes: Term[];
		names: {
			[id: number]: string;
		};
		minRepeatTerm: number;
		maxTerm: number;
	} {
		for (let rule of rules) rule.name.rules.push(rule);

		this.terms = this.terms.filter(
			(t) => t.terminal || t.preserve || rules.some((r) => r.name == t || r.parts.includes(t)),
		);

		let names: { [id: number]: string } = {};
		let nodeTypes = [this.error];

		this.error.id = T.Err;
		let nextID = T.Err + 1;

		// Assign ids to terms that represent node types
		for (let term of this.terms)
			if (term.id < 0 && term.nodeType && !term.repeated) {
				term.id = nextID++;
				nodeTypes.push(term);
			}
		// Put all repeated terms after the regular node types
		let minRepeatTerm = nextID;
		for (let term of this.terms)
			if (term.repeated) {
				term.id = nextID++;
				nodeTypes.push(term);
			}
		// Then comes the EOF term
		this.eof.id = nextID++;
		// And then the remaining (non-node, non-repeat) terms.
		for (let term of this.terms) {
			if (term.id < 0) term.id = nextID++;
			if (term.name) names[term.id] = term.name;
		}
		if (nextID >= 0xfffe) throw new GenError("Too many terms");

		return { nodeTypes, names, minRepeatTerm, maxTerm: nextID - 1 };
	}
}

export function cmpSet<T>(a: readonly T[], b: readonly T[], cmp: (a: T, b: T) => number): number {
	if (a.length != b.length) return a.length - b.length;
	for (let i = 0; i < a.length; i++) {
		let diff = cmp(a[i], b[i]);
		if (diff) return diff;
	}
	return 0;
}

const none: readonly any[] = [];

export class Conflicts {
	readonly precedence: number;
	readonly ambigGroups: readonly string[];
	readonly cut: number;

	constructor(precedence: number, ambigGroups: readonly string[] = none, cut = 0) {
		this.precedence = precedence;
		this.ambigGroups = ambigGroups;
		this.cut = cut;
	}

	join(other: Conflicts): Conflicts {
		if (this == Conflicts.none || this == other) return other;
		if (other == Conflicts.none) return this;
		return new Conflicts(
			Math.max(this.precedence, other.precedence),
			union(this.ambigGroups, other.ambigGroups),
			Math.max(this.cut, other.cut),
		);
	}

	cmp(other: Conflicts): number {
		return (
			this.precedence - other.precedence ||
			cmpSet(this.ambigGroups, other.ambigGroups, (a, b) => (a < b ? -1 : a > b ? 1 : 0)) ||
			this.cut - other.cut
		);
	}

	static none: Conflicts = new Conflicts(0);
}

export function union<T>(a: readonly T[], b: readonly T[]): readonly T[] {
	if (a.length == 0 || a == b) return b;
	if (b.length == 0) return a;
	let result = a.slice();
	for (let value of b) if (!a.includes(value)) result.push(value);
	return result.sort();
}

let ruleID = 0;

export class Rule {
	id: number = ruleID++;

	readonly name: Term;
	readonly parts: readonly Term[];
	readonly conflicts: readonly Conflicts[];
	readonly skip: Term;

	constructor(name: Term, parts: readonly Term[], conflicts: readonly Conflicts[], skip: Term) {
		this.name = name;
		this.parts = parts;
		this.conflicts = conflicts;
		this.skip = skip;
	}

	cmp(rule: Rule): number {
		return this.id - rule.id;
	}

	cmpNoName(rule: Rule): number {
		return (
			this.parts.length - rule.parts.length ||
			this.skip.hash - rule.skip.hash ||
			this.parts.reduce((r, s, i) => r || s.cmp(rule.parts[i]), 0) ||
			cmpSet(this.conflicts, rule.conflicts, (a, b) => a.cmp(b))
		);
	}

	toString(): string {
		return this.name + " -> " + this.parts.join(" ");
	}

	get isRepeatWrap(): boolean {
		return this.name.repeated && this.parts.length == 2 && this.parts[0] == this.name;
	}

	sameReduce(other: Rule): boolean {
		return (
			this.name == other.name &&
			this.parts.length == other.parts.length &&
			this.isRepeatWrap == other.isRepeatWrap
		);
	}
}

File: dotnet/Rezel/Generator/Grammar.cs
using LrTerm = Rezel.Lr.Term;

namespace Rezel.Generator;

public static class TermFlag
{
    public const int Terminal = 1;
    public const int Top = 2;
    public const int Eof = 4;
    public const int Preserve = 8;
    public const int Repeated = 16;
    public const int Inline = 32;
}

public delegate int CmpFunc<in T>(T a, T b);

public static class GrammarUtils
{
    public static bool HasProps(Dictionary<string, string> props)
    {
        return props.Count > 0;
    }

    public static int CmpSet<T>(IReadOnlyList<T> a, IReadOnlyList<T> b, CmpFunc<T> cmp)
    {
        if (a.Count != b.Count) return a.Count - b.Count;
        for (var i = 0; i < a.Count; i++)
        {
            var diff = cmp(a[i], b[i]);
            if (diff != 0) return diff;
        }
        return 0;
    }

    public static List<T> Union<T>(List<T> a, List<T> b) where T : IComparable<T>
    {
        if (a.Count == 0 || ReferenceEquals(a, b)) return b;
        if (b.Count == 0) return a;
        var result = new List<T>(a);
        foreach (var value in b)
            if (!a.Contains(value))
                result.Add(value);
        result.Sort();
        return result;
    }
}

public sealed class Term
{
    private static int _termHash;

    public readonly int Hash = ++_termHash;
    public int Id = -1;
    public List<Rule> Rules = [];

    public readonly string Name;
    private int _flags;
    public readonly string? NodeName;
    public readonly Dictionary<string, string> Props;

    public Term(string name, int flags, string? nodeName, Dictionary<string, string>? props = null)
    {
        Name = name;
        _flags = flags;
        NodeName = nodeName;
        Props = props ?? new Dictionary<string, string>();
    }

    public override string ToString() => Name;

    public bool NodeType => Top || NodeName != null || GrammarUtils.HasProps(Props) || Repeated;

    public bool Terminal => (_flags & TermFlag.Terminal) > 0;
    public bool Eof => (_flags & TermFlag.Eof) > 0;
    public bool Error => Props.ContainsKey("error");
    public bool Top => (_flags & TermFlag.Top) > 0;
    public bool Interesting => _flags > 0 || NodeName != null;
    public bool Repeated => (_flags & TermFlag.Repeated) > 0;

    public bool Preserve
    {
        get => (_flags & TermFlag.Preserve) > 0;
        set => _flags = value ? _flags | TermFlag.Preserve : _flags & ~TermFlag.Preserve;
    }

    public bool Inline
    {
        get => (_flags & TermFlag.Inline) > 0;
        set => _flags = value ? _flags | TermFlag.Inline : _flags & ~TermFlag.Inline;
    }

    public int Cmp(Term other) => Hash - other.Hash;
}

public sealed class TermSet
{
    public List<Term> Terms = [];
    public Dictionary<string, Term> Names = [];
    public readonly Term Eof;
    public readonly Term Error;
    public List<Term> Tops = [];

    public TermSet()
    {
        Eof = Term("\u2404", null, TermFlag.Terminal | TermFlag.Eof);
        Error = Term("\u26A0", "\u26A0", TermFlag.Preserve);
    }

    public Term Term(string name, string? nodeName, int flags = 0, Dictionary<string, string>? props = null)
    {
        var term = new Term(name, flags, nodeName, props);
        Terms.Add(term);
        Names[name] = term;
        return term;
    }

    public Term MakeTop(string? nodeName, Dictionary<string, string> props)
    {
        var term = Term("@top", nodeName, TermFlag.Top, props);
        Tops.Add(term);
        return term;
    }

    public Term MakeTerminal(string name, string? nodeName, Dictionary<string, string>? props = null)
    {
        return Term(name, nodeName, TermFlag.Terminal, props);
    }

    public Term MakeNonTerminal(string name, string? nodeName, Dictionary<string, string>? props = null)
    {
        return Term(name, nodeName, 0, props);
    }

    public Term MakeRepeat(string name)
    {
        return Term(name, null, TermFlag.Repeated);
    }

    public string UniqueName(string name)
    {
        for (var i = 0; ; i++)
        {
            var cur = i > 0 ? $"{name}-{i}" : name;
            if (!Names.ContainsKey(cur)) return cur;
        }
    }

    public sealed class FinishResult
    {
        public required List<Term> NodeTypes;
        public required Dictionary<int, string> Names;
        public required int MinRepeatTerm;
        public required int MaxTerm;
    }

    public FinishResult Finish(Rule[] rules)
    {
        foreach (var rule in rules)
            rule.Name.Rules.Add(rule);

        Terms = Terms.Where(t =>
            t.Terminal || t.Preserve || rules.Any(r => r.Name == t || r.Parts.Contains(t))
        ).ToList();

        var names = new Dictionary<int, string>();
        var nodeTypes = new List<Term> { Error };

        Error.Id = LrTerm.Err;
        var nextID = LrTerm.Err + 1;

        foreach (var term in Terms)
            if (term.Id < 0 && term.NodeType && !term.Repeated)
            {
                term.Id = nextID++;
                nodeTypes.Add(term);
            }

        var minRepeatTerm = nextID;
        foreach (var term in Terms)
            if (term.Repeated)
            {
                term.Id = nextID++;
                nodeTypes.Add(term);
            }

        Eof.Id = nextID++;
        foreach (var term in Terms)
        {
            if (term.Id < 0) term.Id = nextID++;
            if (term.Name != null) names[term.Id] = term.Name;
        }

        if (nextID >= 0xfffe) throw new GenError("Too many terms");

        return new FinishResult
        {
            NodeTypes = nodeTypes,
            Names = names,
            MinRepeatTerm = minRepeatTerm,
            MaxTerm = nextID - 1
        };
    }
}

public sealed class Conflicts
{
    public readonly int Precedence;
    public readonly List<string> AmbigGroups;
    public readonly int Cut;

    public Conflicts(int precedence, List<string>? ambigGroups = null, int cut = 0)
    {
        Precedence = precedence;
        AmbigGroups = ambigGroups ?? [];
        Cut = cut;
    }

    public Conflicts Join(Conflicts other)
    {
        if (ReferenceEquals(this, None) || ReferenceEquals(this, other)) return other;
        if (ReferenceEquals(other, None)) return this;
        return new Conflicts(
            Math.Max(Precedence, other.Precedence),
            GrammarUtils.Union(AmbigGroups, other.AmbigGroups),
            Math.Max(Cut, other.Cut)
        );
    }

    public int Cmp(Conflicts other)
    {
        return (Precedence - other.Precedence) != 0
            ? Precedence - other.Precedence
            : GrammarUtils.CmpSet(AmbigGroups, other.AmbigGroups, (a, b) => string.CompareOrdinal(a, b)) != 0
                ? GrammarUtils.CmpSet(AmbigGroups, other.AmbigGroups, (a, b) => string.CompareOrdinal(a, b))
                : Cut - other.Cut;
    }

    public static readonly Conflicts None = new Conflicts(0);
}

public sealed class Rule
{
    private static int _ruleID;

    public readonly int Id = _ruleID++;
    public readonly Term Name;
    public readonly Term[] Parts;
    public readonly Conflicts[] Conflicts;
    public readonly Term Skip;

    public Rule(Term name, Term[] parts, Conflicts[] conflicts, Term skip)
    {
        Name = name;
        Parts = parts;
        Conflicts = conflicts;
        Skip = skip;
    }

    public int Cmp(Rule rule) => Id - rule.Id;

    public int CmpNoName(Rule rule)
    {
        var lenDiff = Parts.Length - rule.Parts.Length;
        if (lenDiff != 0) return lenDiff;
        var skipDiff = Skip.Hash - rule.Skip.Hash;
        if (skipDiff != 0) return skipDiff;
        for (var i = 0; i < Parts.Length; i++)
        {
            var partDiff = Parts[i].Cmp(rule.Parts[i]);
            if (partDiff != 0) return partDiff;
        }
        return GrammarUtils.CmpSet(Conflicts, rule.Conflicts, (a, b) => a.Cmp(b));
    }

    public override string ToString() => Name + " -> " + string.Join(" ", (object[])Parts);

    public bool IsRepeatWrap => Name.Repeated && Parts.Length == 2 && Parts[0] == Name;

    public bool SameReduce(Rule other)
    {
        return Name == other.Name &&
               Parts.Length == other.Parts.Length &&
               IsRepeatWrap == other.IsRepeatWrap;
    }
}
